src/app/Shared/Services/Api.service.tsx

```diff
@@ -71,7 +70,8 @@ export class ApiService {
 
   constructor(
```

Comment:
I'm trying to find the reason why my branch makes a call to `v1/targets` and `v1/recordings` on the login page and why the HTTP responses appear as toast notifications. On the main branch login page, there are two `401` responses that don't appear as toast notifications. 

I see that `queryForTargets()` gets called in the `TargetsService` constructor and `<TargetSelect/>`.  `doGet<Recording[]>()` is also called in the `ApiService` constructor and a few other components. Since`<TargetSelect />` isn't rendered on the login page, why would I see a call to `/v1/targets` on my branch?

I also noticed the `notifications` request results in a `101 protocol switched` - which protocol is this referring to? Is this relevant to this problem?
![image](https://user-images.githubusercontent.com/84587295/131739107-5d081986-3e06-4f3d-b25a-18d12795643d.png)
![image](https://user-images.githubusercontent.com/84587295/131739651-b4d0296d-b763-4a5e-a91d-dc3f5ec4e6f6.png)

Reply:
`101 protocol switched` looks like it's just this: https://developer.mozilla.org/en-US/docs/Web/HTTP/Status/101
It's a normal part of the WebSocket connection procedure.

HTTP error responses are automatically turned into notifications by the `Api.service` through certain call paths. It looks like you're seeing it coming from here: https://github.com/cryostatio/cryostat-web/blob/main/src/app/Shared/Services/Api.service.tsx#L548 - many of the public methods in that class pipe their results like `.pipe(handleError(err => this.handleError<Response>(err, req))`, so the original component that subscribed to this request Observable will observe the error event, and a notification will be displayed automatically.

Other HTTP calls that don't go through this call path, like the `checkAuth` function (https://github.com/cryostatio/cryostat-web/blob/a29e59b1cf7caed9416722bdcccfd6814fb3a3d9/src/app/Shared/Services/Api.service.tsx#L127), won't automatically display the toast notification. In this particular case it's mostly because of that startup "probing" `/auth` request that I described earlier - that happens every time the React app is opened, and in any real deployment it will fail with a 401, so it would be annoying/confusing to show a notification here. Notifications are instead only shown when the user explicitly tries to authenticate and *that* fails (https://github.com/cryostatio/cryostat-web/blob/main/src/app/Login/Login.tsx#L70).


> I see that queryForTargets() gets called in the TargetsService constructor and <TargetSelect/>. doGet<Recording[]>() is also called in the ApiService constructor and a few other components. Since<TargetSelect /> isn't rendered on the login page, why would I see a call to /v1/targets on my branch?

Without analyzing your code too deeply (yet), I would suspect that the application is briefly considering the user as authenticated and renders the Dashboard view, even just for a single frame, which does contain a TargetSelect component and would trigger the `GET /api/v1/targets` request. Then, state is properly restored and the app considers the user not-authenticated, kicking them back to the Login page and removing the TargetSelect/Dashboard from the rendered view, but not before the request is fired.

Reply:
I tried setting the initial value of  `ReplaySubject authenticated` and the state variable in `routes.tsx` based on the contents of sessionStorage. The `useEffect` hook is also called every time the browser refreshes. The web page still flickers between the login page and the dashboard when I try to login or refresh the page. What else could cause the application to briefly read `authenticated` as `true`?

`routes.tsx`:
```
const [authenticated, setAuthenticated] = React.useState(context.login.isAuthenticated());

  React.useEffect(() => {
    const sub = combineLatest(context.login.getAuthenticated().pipe(distinctUntilChanged()),
      context.notificationChannel.isReady().pipe(distinctUntilChanged()))
    .subscribe(
      ([isAuthenticated, notificationChannelReady]) => {
        setAuthenticated(isAuthenticated && notificationChannelReady);
      },
      err => console.error(err),
    );
    return () => sub.unsubscribe();
}, [context.login, context.notificationChannel, addSubscription, setAuthenticated]);
```
`LoginService`:
```
 constructor() {
    ...
    this.token.next(this.replaceWithCachedToken(''));
    this.authenticated.next(this.isAuthenticated());
  }
...
  private replaceWithCachedToken(defaultToken: string) {
    const cachedToken = this.getCachedToken();
    return (!!cachedToken) ? cachedToken : defaultToken;
  }
...
  isAuthenticated() : boolean {
    return !!this.getCachedToken();
  }
...
  private getCachedToken() : string {
    const token = sessionStorage.getItem('token');
    return (!!token) ? token : '';
  }
```

Reply:
Update: I'm going to try to change the way `LoginService` emits changes to `authenticated`. `loginService` can emit one event when the user logs in and one event when the user logs out. Then `routes` and `AppLayout` can subscribe to those messages instead of relying on the value of the replay subject `authenticated` at that moment.